### PR TITLE
fixes for strided conv and intermediate norm

### PR DESCRIPTION
**Pull request by Pagey**
hey @Bartzi 
1. added support for strided conv using adj and Crop
2. (without connection to fix 1) added intermediate normalizations for feature maps- it isn't stated in original paper but was commented by bojarski comment from 16/11/17 on https://github.com/mbojarski/VisualBackProp/issues/1

### CODE REVIEW

File: insights/visual_backprop.py [hunk 1]
```diff
@@ -11,10 +11,27 @@ def combine_feature_maps(intermediate_symbols, node, nodes, scaled_feature):
     data_input_node = nodes[node['inputs'][0][0]]
     try:
         data_input_symbol = intermediate_symbols["{}_output".format(data_input_node['name'])]
+#    except ValueError:
```
Comment (Bartzi) [inline insights/visual_backprop.py]: could you remove those comments?

File: insights/visual_backprop.py [hunk 2]
```diff
@@ -11,10 +11,27 @@ def combine_feature_maps(intermediate_symbols, node, nodes, scaled_feature):
...
+#        data_input_symbol = intermediate_symbols[data_input_node['name']]
     except ValueError:
-        data_input_symbol = intermediate_symbols[data_input_node['name']]
+        try:
+            data_input_symbol = intermediate_symbols[data_input_node['name']]
+        except ValueError:
+            data_input_symbol = intermediate_symbols["{}_output0".format(data_input_node['name'])] # MY looks like lrn has output0
```
Comment (Bartzi) [inline insights/visual_backprop.py:26]: this seems to be specific for your usecase, isn't it? I mean the symbol could get another name and then this code will break.

File: insights/visual_backprop.py [hunk 3]
```diff
@@ -11,10 +11,27 @@ def combine_feature_maps(intermediate_symbols, node, nodes, scaled_feature):
...
+
+
     averaged_feature_map = mx.symbol.mean(data_input_symbol, axis=1, keepdims=True)
-    feature_map = scaled_feature * averaged_feature_map
+	
+    # MY normalization test- by bojarski comment from 16/11/17 on https://github.com/mbojarski/VisualBackProp/issues/1
+    min_value = mx.symbol.min(averaged_feature_map)
```
Comment (Bartzi) [inline insights/visual_backprop.py]: could you refactor this into a method? This will remove code duplication

File: insights/visual_backprop.py [hunk 4]
```diff
@@ -27,39 +44,55 @@ def build_visual_backprop_symbol(start_symbol, input_name=None):
...
+    min_value = mx.symbol.min(feature_map)
+    max_value = mx.symbol.max(feature_map)
+    feature_map = mx.symbol.broadcast_sub(feature_map, min_value)
+    feature_map = mx.symbol.broadcast_mul(feature_map, 1.0 / (max_value - min_value))
+    # end MY normalization test
+
     for node in reversed(nodes):
         if input_name is not None and node['name'] == input_name or node['name'] == 'data':
-            break
+            breakPagey
```
Comment (Bartzi) [inline insights/visual_backprop.py]: is there a typo in here?

File: insights/visual_backprop.py [hunk 5]
```diff
@@ -27,39 +44,55 @@ def build_visual_backprop_symbol(start_symbol, input_name=None):
...
+                adj=(stride_height - 1,
+                     stride_width - 1),  # MY this is sometimes needed- will take care of it with Crop if needed in combine_feature_maps
                 num_filter=1,
             )
             feature_map = combine_feature_maps(intermediate_symbols, node, nodes, scaled_feature)
         else:
             continue
 
-    # normalize feature map
+    # normalize feature map - (MY this isn't usually really necessary if we normalize in combine_feature_maps, but shouldn't hurt)
```
Comment (Bartzi) [inline insights/visual_backprop.py]: if we have a normalization function we should call it here, or just leave this part out completely.

File: insights/visual_backprop.py [hunk 6]
```diff
@@ -27,39 +44,55 @@ def build_visual_backprop_symbol(start_symbol, input_name=None):
...
     min_value = mx.symbol.min(feature_map)
     max_value = mx.symbol.max(feature_map)
     feature_map = mx.symbol.broadcast_sub(feature_map, min_value)
     feature_map = mx.symbol.broadcast_mul(feature_map, 1.0 / (max_value - min_value))
-    return feature_map
+    return feature_map
```
Comment (Bartzi) [inline insights/visual_backprop.py]: please add a newline to the end of the file

**Review by Bartzi — CHANGES_REQUESTED** [review]
Thanks, but please have a look at my comments ;)

File: insights/visual_backprop.py (hunk 2 above)
Comment (Pagey) [reply insights/visual_backprop.py:26]: nope this seems to be the mxnet coded node name for LRN (which isn't really used much more, but is still part of e.g. AlexNet)- your generic code is the first case of the handler, this is only used if that one isn't found (as in the case of LRN)

**Review by Bartzi — APPROVED** [review]

**Comment by Bartzi** [comment]
alright, thank you very much!

**Pull request merged**